nflow-tests/src/main/java/io/nflow/tests/demo/workflow/BulkWorkflow.java

```diff
@@ -0,0 +1,149 @@
+  public static final String VAR_CONCURRENCY = "concurrency";
+
+  private static final EnumSet<WorkflowInstanceStatus> RUNNING_STATES = complementOf(EnumSet.of(finished, created));
+  private static final Logger logger = getLogger(BulkWorkflow.class);
+
+  @Inject
+  WorkflowInstanceService instanceService;
+
+  public enum State implements io.nflow.engine.workflow.definition.WorkflowState {
+    splitWork(start), waitForChildrenToFinish(normal), done(end), error(manual);
+
+    private WorkflowStateType type;
+
+    State(WorkflowStateType type) {
+      this.type = type;
+    }
+
+    @Override
+    public WorkflowStateType getType() {
+      return type;
+    }
+
+    @Override
+    public String getDescription() {
+      return name();
+    }
+  }
+
+  protected BulkWorkflow(String type) {
+    super(type, splitWork, error);
+    setDescription("Executes child workflows in bulk but gracefully without effecting non-bulk tasks.");
+    permit(splitWork, waitForChildrenToFinish);
+    permit(waitForChildrenToFinish, done);
+  }
+
+  public BulkWorkflow() {
+    this(BULK_WORKFLOW_TYPE);
+  }
+
+  public NextAction splitWork(StateExecution execution, @StateVar(value = VAR_CHILD_DATA, readOnly = true) JsonNode data) {
+    boolean childrenFound = splitWorkImpl(execution, data);
+    if (childrenFound) {
+      return moveToState(waitForChildrenToFinish, "Running");
+    }
+    return retryAfter(waitForChildrenUntil(), "Waiting for child workflows");
+  }
+
+  protected boolean splitWorkImpl(StateExecution execution, @SuppressWarnings("unused") JsonNode data) {
+    if (execution.getAllChildWorkflows().isEmpty()) {
+      throw new RuntimeException(
+          "No child workflows found - either add them before starting the parent or implement splitWorkflowImpl");
+    }
+    return true;
+  }
+
+  protected DateTime waitForChildrenUntil() {
+    return now().plusMinutes(1);
+  }
+
+  public NextAction waitForChildrenToFinish(StateExecution execution,
+      @StateVar(value = VAR_CONCURRENCY, readOnly = true) int concurrency) {
+    List<WorkflowInstance> childWorkflows = execution.getAllChildWorkflows();
+    long completed = 0;
+    long running = 0;
+    for (WorkflowInstance child : childWorkflows) {
+      if (child.status == finished) {
+        completed++;
+      } else if (isRunning(child)) {
+        running++;
+      }
+    }
+    if (completed == childWorkflows.size()) {
+      return moveToState(done, "All children completed");
+    }
+    long toStart = min(max(1, concurrency) - running, childWorkflows.size() - completed);
+    if (toStart > 0) {
+      childWorkflows.stream().filter(this::isInInitialState).limit(toStart).forEach(this::wakeup);
+      logger.info("Started " + toStart + " child workflows");
+    }
+    long progress = completed * 100 / childWorkflows.size();
+    return retryAfter(waitForChildrenToCompleteUntil(), "Waiting for child workflows to complete - " + progress + "% done");
```

Comment:
Did we have a limit on number of retries? This polling could be retried many times if there are many child workflows. Should we increase the limit or add some flag to make this retry not increase the counter.

Reply:
Changed retry limit for bulk workflow to Integer.MAX_VALUE. Is this ok or should we use moveToNextStateAfter instead?